atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java

```diff
@@ -110,60 +114,79 @@ public void run() {
             Thread.sleep(getBackoffTimeWhenSweepHasNotRun());
             log.info("Starting background sweeper.");
             while (true) {
-                long millisToSleep = checkConfigAndRunSweep(locks);
-                Thread.sleep(millisToSleep);
+                SweepOutcome outcome = checkConfigAndRunSweep(locks);
+
+                updateBatchSize(outcome);
+                updateMetrics(outcome);
+
+                sleepUntilNextRun(outcome);
             }
         } catch (InterruptedException e) {
             log.warn("Shutting down background sweeper. Please restart the service to rerun background sweep.");
         }
     }
 
-    // Returns milliseconds to sleep
+    private void updateBatchSize(SweepOutcome outcome) {
+        if (outcome == SweepOutcome.SUCCESS) {
+            batchSizeMultiplier = Math.min(1.0, batchSizeMultiplier * 1.01);
+            return;
+        }
+        if (outcome == SweepOutcome.RETRYING_WITH_SMALLER_BATCH) {
```

Comment:
What if the outcome is ERROR?

Reply:
I think with the ERROR outcome something more serious is wrong than just the batch size - so this is unlikely to help.  But to avoid an unnecessary behaviour change I'll also shrink the batch on ERROR.